### ISSUE TITLE
"Table 'jhipster.jhi_user' doesn't exist" error on new app

### ISSUE THREAD

**Issue by piedrahitac** [jhipster/generator-jhipster#2018]
## Overview

After creating a new app using MySQL and OAuth2 on Linux, the app is inaccessible because the user always receives an authentication error.

The logs report `Table 'jhipster.jhi_user' doesn't exist` even though the table was created correctly upon application deployment.

The problem seems to be associated with the case of the table name. The query is using lowercase while the table was created using uppercase. Using the same case in the query and the table name seems to fix the issue.

This didn't happen in previous versions of JHipster.

## JHipster version(s)

2.21.0

## JHipster configuration
```json
`{
  "generator-jhipster": {
    "baseName": "jhipster",
    "packageName": "com.piedrahitac.jhipster",
    "packageFolder": "com/piedrahitac/jhipster",
    "authenticationType": "oauth2",
    "hibernateCache": "ehcache",
    "clusteredHttpSession": "no",
    "websocket": "no",
    "databaseType": "sql",
    "devDatabaseType": "mysql",
    "prodDatabaseType": "mysql",
    "searchEngine": "no",
    "useSass": true,
    "buildTool": "maven",
    "frontendBuilder": "grunt",
    "javaVersion": "8",
    "enableTranslation": true,
    "rememberMeKey": "obscured"
  }
}`
```
## Browsers and OS

Chrome and Firefox on Linux (Fedora 22)

## Reproduce the error

Create a brand new JHipster app, attempt to login and check the log.

**Comment by deepu105** [comment]
@jdubois @andidev @atomfrede @gmarziou can any one try this im unable to reproduce this

**Comment by atomfrede** [comment]
Will try it out.

**Comment by gmarziou** [comment]
I confirm the bug with MySql, it works fine with H2 and PosgreSQL.
I tested on Ubuntu, it could work on Windows according to this doc:

https://dev.mysql.com/doc/refman/5.0/en/identifier-case-sensitivity.html

**Comment by deepu105** [comment]
That explains why i couldn't reproduce it coz im on windows 10

**Comment by gmarziou** [comment]
I got it working by switching back from org.springframework.boot.orm.jpa.hibernate.SpringNamingStrategy  to org.hibernate.cfg.EJB3NamingStrategy in application*.yml. It's probably not the final solution.

I am intrigued by the oauth2 tables being in lower case while the JHI ones are in upper case with both naming strategies.

![screenshot from 2015-09-18 17 16 17](https://cloud.githubusercontent.com/assets/361007/9963180/1f90b33c-5e2a-11e5-80b4-552a53812e43.png)

**Comment by jdubois** [comment]
- For the oauth tables in lower-case: this is clearly an issue, as we should be consistent -> I'll refactor this ASAP
- For the MySQL issue: I've already seen this reported. I thought that the case wasn't important for database tables, it turns out that there is an option for that on MySQL for Linux... And that depending on your installation you need to have the right case.

-> we need to find again this option, and document it clearly. Maybe even test for this when JHipster starts!

**Comment by jdubois** [comment]
Sorry, my previous comment was wrong for the oauth tables!!

The tables are named like this because this is how they are implemented by the JdbcTokenStore from Spring Security. We have the default table names, as per our [Policy 1](http://jhipster.github.io/policies.html). So this perfectly normal, and should not be changed.

**Comment by gmarziou** [comment]
> For the MySQL issue: I've already seen this reported. I thought that the case wasn't important for database tables, it turns out that there is an option for that on MySQL for Linux... And that depending on your installation you need to have the right case.

Yes this is `lower_case_table_names` system variable as explained [in the official doc](https://dev.mysql.com/doc/refman/5.0/en/identifier-case-sensitivity.html).
The problem is that the default value is not the same for Linux and OSX.

So I guess we could execute a simple select in DatabaseConfiguration and depending on result log a diagnostic and advice. I suspect that not everybody could change this value, especially on shared hosting as it is a global setting.

What if we were using EJB3NamingStrategy for MySQL? Would it cause any issue somewhere?

**Comment by deepu105** [comment]
@gmarziou i refactored the table names in last release to be consistent ie snake_case every where and changed naming strategy i guess if we have to workaround for mySQL we need to take care not to break join tables and foreign key names which are handled different by both strategy 
Is there any way to force table name case in liquibase or something?

**Comment by deepu105** [comment]
Or we can make table names lowercase by default and it shouldn't affect other os and database right?

**Comment by gmarziou** [comment]
When you quote table names at creation, the names are respected as provided.
Liquibase has an option for this: `objectQuotingStrategy="QUOTE_ALL_OBJECTS"`  that can be used as an attribute of `databaseChangeLog` or `changeSet` tags. I haven't tested it.

Having all table names lower case would probably be safe on all OS while bringing some consistency with oauth tables, I don't know if we have other table naming  brought by other frameworks than Spring Security.

Anyway, I'm not sure we can avoid breaking changes.

**Comment by robertolos** [comment]
I temporary fixed the issue changing the mysql configuration to be table name case insensitive:

Edit /etc/mysql/my.cnf
Underneath the [mysqld] section add:
    ```lower_case_table_names = 1```
Restart mysql

**Comment by gmarziou** [comment]
I got it working by converting to lower case all table names in 0000*_initial_schema.xml and JPA entities and adding `objectQuotingStrategy="QUOTE_ALL_OBJECTS"` to databaseChangelog root tag in this file. 

I could make a PR if you think this solution is acceptable.

**Comment by jdubois** [comment]
I've always liked my tables to be uppercase, but I guess this is just a personal preference.

If nobody is against it, and if it does solve this annoying issue, yes let's do it that way!

**Comment by gmarziou** [comment]
So far, I made the change only for JHipster tables, should I do it also for entity generator?

**Comment by gmarziou** [comment]
I had to add `;DATABASE_TO_UPPER=FALSE` to JDBC URL for H2 to make it respect lower case table names and it seems that Oracle would behave the same.

**Comment by sammubr** [comment]
This `;DATABASE_TO_UPPER=FALSE` causes:

`[ERROR] org.hibernate.tool.hbm2ddl.SchemaValidator - HHH000319: Could not get database metadata
org.h2.jdbc.JdbcSQLException: Schema "information_schema" not found; SQL statement:
select sequence_name from information_schema.sequences [90079-187]`

when running tests.

**Comment by gmarziou** [comment]
Right but this can be worked around by modifying the FixedH2Dialect by adding:

~~~java
    @Override
    public String getQuerySequencesString() {
        return "select sequence_name from INFORMATION_SCHEMA.sequences";
    }
~~~

However, there are other issues raising after.
So maybe it's not worth to dig further in this direction and t(he safest could be to advice MySQL users to use `lower_case_table_names = 1`

**Comment by deepu105** [comment]
@gmarziou what happens if we just make the Table names lowercase in liquibase and @Table annotation?

Oracle shouldnt have issues as it will read any case and I guess H2 works similar way but not sure about mySQL and Posgres

**Comment by gmarziou** [comment]
If we do so, we have a problem with H2.

**Comment by jdubois** [comment]
I'm not a big fan of the `lower_case_table_names = 1` option, as people might not be able to change this option in their database (if it's hosted by some other entity, for example).
Is this really the only good option available? @gmarziou did you work on the DatabaseConfiguration, at least to throw a log comment when there is this issue?

**Comment by gmarziou** [comment]
I think the solution is to use lowercase table names without setting objectQuotingStrategy="QUOTE_ALL_OBJECTS" in Liquibase, I'll do more testing tomorrow.

**Comment by jdubois** [comment]
OK @gmarziou I believe you are right. It's the only solution where it works "out of the box" for people, so it's better.

**Comment by gmarziou** [comment]
This could be a breaking change for existing apps

**Comment by coolduebtn** [comment]
Hello guys, I just started with jhipster and I have the same problem...i am stuck because of this. your commit to make all tables lower case will be available from when to use ??

**Comment by robertolos** [comment]
The bug has been fixed in the last release 2.22.0. Check this page for the list of bug fixed and the instructions to upgrade jhipster http://jhipster.github.io/2015/10/06/jhipster-release-2.22.0.html

**Comment by coolduebtn** [comment]
Thanks a lot for the update..its working now... i am really excited about jhipster :)
